crypto/src/main/java/org/web3j/crypto/StructuredDataEncoder.java

```diff
@@ -97,9 +106,106 @@ public static String encodeType(String primaryType) {
+        List<Integer> dimensions = new ArrayList<>();
+        while (dimensionTypeMatcher.find()) {
+            String currentDimension = dimensionTypeMatcher.group(1);
+            if (currentDimension == null) {
+                dimensions.add(Integer.parseInt("-1"));
+            } else {
+                dimensions.add(Integer.parseInt(currentDimension));
+            }
+        }
+
+        return dimensions;
+    }
+
+    public static ArrayList<Pair> getDepthsAndDimensions(Object data, int depth) {
+        if (!ArrayList.class.isInstance(data)) {
+            // Nothing more to recurse, since the data is no more an array
+            return new ArrayList<>();
+        }
+
+        ArrayList<Pair> list = new ArrayList<>();
+        ArrayList<Object> dataAsArray = (ArrayList<Object>) data;
+        list.add(new Pair(depth, dataAsArray.size()));
+        for (Object subdimensionalData: dataAsArray) {
+            list.addAll(getDepthsAndDimensions(subdimensionalData,depth + 1));
+        }
+
+        return list;
+    }
+
+    public static ArrayList<Integer> getArrayDimensionsFromData(
+            Object data) throws ValidationException {
+        ArrayList<Pair> depthsAndDimensions = getDepthsAndDimensions(data, 0);
+        // groupedByDepth has key as depth and value as ArrayList(pair(Depth, Dimension))
+        Map<Object, List<Pair>> groupedByDepth = depthsAndDimensions.stream().collect(
+                Collectors.groupingBy(
+                        depthDimensionPair -> depthDimensionPair.getKey()
+                )
+        );
+
+        // depthDimensionsMap is aimed to have key as depth and value as ArrayList(Dimension)
+        Map<Integer, List<Integer>> depthDimensionsMap = new HashMap<>();
+        for (Map.Entry<Object, List<Pair>> entry : groupedByDepth.entrySet()) {
+            List<Integer> pureDimensions = new ArrayList<>();
+            for (Pair depthDimensionPair: entry.getValue()) {
+                pureDimensions.add((Integer) depthDimensionPair.getValue());
+            }
+            depthDimensionsMap.put((Integer) entry.getKey(), pureDimensions);
+        }
+
+        ArrayList<Integer> dimensions = new ArrayList<>();
+        for (Map.Entry<Integer, List<Integer>> entry : depthDimensionsMap.entrySet()) {
+            Set<Integer> setOfDimensionsInParticularDepth = new TreeSet<>(entry.getValue());
+            if (setOfDimensionsInParticularDepth.size() != 1) {
+                throw new ValidationException(
+                        String.format(
+                                "Depth %d of array data has more than one dimensions",
+                                entry.getKey()
+                        )
+                );
+            }
+            dimensions.add(setOfDimensionsInParticularDepth.stream().findFirst().get());
+        }
+
+        return dimensions;
+    }
+
+    public static ArrayList<Object> flattenMultidimensionalArray(Object data) {
+        if (!ArrayList.class.isInstance(data)) {
```

Comment:
What if it's not an ArrayList but, say, LinkedList? Wouldn't a more generic List be better here? Or even Iterable?